smart-contract/assembly/contracts/deweb-interface.ts

```diff
@@ -1,187 +1,326 @@
-import {
-  ChunkPost,
-  ChunkGet,
-  PreStore,
-  ChunkDelete,
-} from './serializable/Chunk';
 
 import { Args, stringToBytes, u32ToBytes } from '@massalabs/as-types';
+import { FileChunkPost } from './serializable/FileChunkPost';
+import { FileDelete } from './serializable/FileDelete';
+import { FileChunkGet } from './serializable/FileChunkGet';
+import { Metadata } from './serializable/Metadata';
 import {
+  _editFileMetadata,
+  _editGlobalMetadata,
+  _getFileMetadata,
+  _getGlobalMetadata,
+  _removeFileMetadata,
+  _removeGlobalMetadata,
+} from './internals/metadata';
+import {
+  _setFileChunk,
   _getFileChunk,
   _getTotalChunk,
-  _setFileChunk,
-  _setTotalChunk,
   _deleteFile,
-  _removeChunksRange,
-  _removeFile,
 } from './internals/chunks';
-
-import { _removeFilePath } from './internals/file-list';
-import { FILES_PATH_LIST } from './internals/const';
-import { _pushFilePath } from './internals/file-list';
+import { _getFileLocations } from './internals/location';
+import { _fileInit } from './internals/fileInit';
+import { FileInit } from './serializable/FileInit';
 
 /**
  * Initializes the smart contract.
- * Sets the contract deployer as the owner and initializes an empty file path list.
- * @param _ - Unused parameter (required).
+ * Sets the contract deployer as the owner.
  */
 export function constructor(_: StaticArray<u8>): void {
   if (!Context.isDeployingContract()) return;
+
   _setOwner(Context.caller().toString());
-  FILES_PATH_LIST.set([]);
 }
 
+/* -------------------------------------------------------------------------- */
+/*                             FILE INITIALIZATION                            */
+/* -------------------------------------------------------------------------- */
+
+/**
+ * Initializes the contract storage with the given files and metadata.
+ * @param _binaryArgs - Serialized arguments containing arrays of FileInit objects,
+ *                      files to delete, and global metadata.
+ * @throws If the files are invalid or if the caller is not the owner.
+ */
+export function filesInit(_binaryArgs: StaticArray<u8>): void {
+  _onlyOwner();
+
+  const args = new Args(_binaryArgs);
+  const files = args
+    .nextSerializableObjectArray<FileInit>()
+    .expect('Invalid files to initialize');
+
+  const filesToDelete = args
+    .nextSerializableObjectArray<FileInit>()
+    .expect('Invalid files to delete');
+
+  const globalMetadata = args
+    .nextSerializableObjectArray<Metadata>()
+    .expect('Invalid global metadata');
+
+  for (let i = 0; i < filesToDelete.length; i++) {
+    _deleteFile(filesToDelete[i].hashLocation);
+  }
+
+  for (let i = 0; i < globalMetadata.length; i++) {
+    _editGlobalMetadata(
+      stringToBytes(globalMetadata[i].key),
+      stringToBytes(globalMetadata[i].value),
+    );
+  }
+
+  for (let i = 0; i < files.length; i++) {
+    _fileInit(
+      files[i].location,
+      files[i].hashLocation,
```

Comment:
do not trust the caller on the hash, compute it (we compute it anyway for an assert in _fileInit)